lib/strzero.c

```diff
@@ -0,0 +1,96 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ * SPDX-License-Identifier: curl
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#include "strzero.h"
+
+/*
+ * Curl_explicit_bzero zeroes out a memory buffer in such a way to ensure that
+ * it won't be optimized away by the compiler due to being a dead-store. If no
+ * API for this exists it will fall back to plain memset using a volatile ptr.
+ */
+
+#ifdef HAVE_MEMSET_S
+void
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  (void)memset_s(buffer, length, 0, length);
+}
+#elif defined(HAVE_BZERO)
+void
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  bzero(buffer, length);
+}
+#elif defined(HAVE_EXPLICIT_BZERO)
+Curl_explicit_bzero(void *buffer, size_t length)
+{
+  if(!buffer)
+    return;
+  explicit_bzero(buffer, length);
+}
+#elif defined(_WIN32)
```

Comment:
Would it make sense to move this to the top of the list and use this native function always, on Windows?

It's an intrinsic function and should work in all versions.

Reply:
Fair point.